src/vs/workbench/contrib/chat/browser/chatPasteProviders.ts

```diff
@@ -89,6 +92,19 @@ export class PasteImageProvider implements DocumentPasteEditProvider {
 
 		widget.attachmentModel.addContext(imageContext);
 
+		this.undoRedoService.pushElement({
```

Comment:
Let's see if we can move this into the `BulkEditService`. `provideDocumentPasteEdits ` should not have any side effects, as this method may be called even if the edit is not actually applied yet

Here's what I'm thinking:

- In our internal version of [`WorkspaceEdit`](https://github.com/microsoft/vscode/blob/4fb64349bcd479e1a0be1a146fedc98264364f0b/src/vs/editor/common/languages.ts#L1754), add a new edit type called something like `ICustomEdit`

- `ICustomEdit ` would be similar to the existing `IWorkspaceUndoRedoElement`. The main methods it would have are  `redo` to apply the edit, and undo to undo it

- Then update the bulk edit service to handle this new type of edit. This is will be the most complicated part as an `ICustomEdit ` would need to work along side the normal text edits

@jrieken Does this sound like a reasonable proposal?

Two alternatives approaches :

- If we'd prefer a more scoped fix, we could instead isolate this logic just to the copy paste controller so that `DocumentPasteEdit` can either return a `WorkspaceEdit` or a custom edit with `undo` and `redo`

- We could add an explicit `AddChatAttachment` edit type and make the bulk edit service know how to handle that

Reply:
I think both make sense. Adding a special workspace edit might be a bit more involed but could be the right thing in the long run (it will enable more undo-able things) but it is also a bit weird that the `DocumentPasteEdit#insertText` is empty then